paderbox/array/sparse.py

```diff
@@ -0,0 +1,701 @@
+    # This is required for pt.data.batch.example_to_device to work
+    _segments: list = field(default_factory=list)
+
+    # Other instance attributes
+    pad_value = 0.
+    _is_torch = None
+    dtype = None
+
+    def __post_init__(self):
+        self.shape = _normalize_shape(self.shape)
+        self._update_dtype(self.dtype)
+
+    @classmethod
+    def from_array_and_offset(cls, array, offset, shape):
+        out = SparseArray(shape=shape)
+        out._add_segment(_SparseSegment(offset, array))
+        return out
+
+    def as_contiguous(self, dtype=None):
+        arr = self._new_full(dtype=dtype)
+        for segment in self._segments:
+            s, array = _clip(arr, segment.array, segment.onset)
+            arr[..., s] = array
+        return arr
+
+    def _update_dtype(self, dtype, pad_value=0.):
+        self.dtype = dtype
+        if dtype is None and self._segments:
+            self.dtype = self._segments[0].array.dtype
+        if self.dtype is not None:
+            self._is_torch = isinstance(self.dtype, torch.dtype)
+            if self._is_torch:
+                self.pad_value = torch.tensor(pad_value, dtype=self.dtype)
+            else:
+                self.pad_value = self.dtype.type(pad_value)
+        if self._segments:
+            assert all(s.array.dtype == self.dtype for s in self._segments), (
+                [s.array.dtype for s in self._segments], self.dtype)
+
+    @property
+    def interval(self) -> ArrayInterval:
+        ai = pb.array.interval.zeros(self.shape[-1])
+        s: _SparseSegment
+        for s in self._segments:
+            ai[s.onset:s.offset] = True
+        return ai
+
+    def _new_full(self, *, shape=None, dtype=None, fill_value=None):
+        if shape is None:
+            shape = self.shape
+        if dtype is None:
+            dtype = self.dtype
+        if fill_value is None:
+            fill_value = self.pad_value
+        if self._is_torch:
+            return torch.full(
+                size=shape, fill_value=fill_value, dtype=dtype
+            )
+        else:
+            return np.full(
+                shape=shape, fill_value=fill_value, dtype=dtype
+            )
+
+    def _add_segment(self, segment: _SparseSegment):
+        if self._segments and (
+                self._is_torch and not isinstance(segment.array, torch.Tensor)
+                or not isinstance(segment.array, np.ndarray)
+        ):
+            raise TypeError(type(segment.array))
+
+        # Check input dtype and shape
+        if self.dtype is None:
+            # This is the first segment we add. It defines dtype and pad value
+            self._update_dtype(segment.array.dtype)
+        elif self.dtype != segment.array.dtype:
+            raise TypeError(
+                f'Type mismatch: SparseArray has dtype {self.dtype}, but '
+                f'assigned array has dtype {segment.array.dtype}'
+            )
+
+        if segment.array.shape[:-1] != self.shape[:-1]:
```

Comment:
Can you add a hint, that the leading dimensions must be equal? Add least as comment in the source code.